lib/utils/oidc.js

```diff
@@ -0,0 +1,138 @@
+         * ```
+         * permissions:
+         *    id-token: write
+         * ```
+         * @see https://docs.github.com/en/actions/security-for-github-actions/security-hardening-your-deployments/configuring-openid-connect-in-cloud-providers#adding-permissions-settings
+         */
+        if (
+          process.env.ACTIONS_ID_TOKEN_REQUEST_URL &&
+          process.env.ACTIONS_ID_TOKEN_REQUEST_TOKEN
+        ) {
+          log.silly('oidc', '"GITHUB_ACTIONS" detected with "ACTIONS_ID_" envs, fetching id_token')
+
+          /**
+           * The specification for an audience is `npm:registry.npmjs.org`,
+           * where "registry.npmjs.org" can be any supported registry.
+           */
+          const audience = `npm:${new URL(registry).hostname}`
+
+          const url = new URL(process.env.ACTIONS_ID_TOKEN_REQUEST_URL)
+          url.searchParams.append('audience', audience)
+          const response = await fetch(url.href, {
+            method: 'POST',
+            retry: opts.retry,
+            headers: {
+              Accept: 'application/json',
+              Authorization: `Bearer ${process.env.ACTIONS_ID_TOKEN_REQUEST_TOKEN}`,
+            },
+          })
+
+          if (!response.ok) {
+            throw new Error(`Failed to fetch id_token from GitHub: received an invalid response`)
+          }
+          const json = await response.json()
+          if (!json.value) {
+            throw new Error(`Failed to fetch id_token from GitHub: missing value`)
+          }
+          log.silly('oidc:', 'GITHUB_ACTIONS valid fetch response for id_token')
+          idToken = json.value
+        } else {
+          throw new Error('GITHUB_ACTIONS detected. If you intend to publish using OIDC, please set workflow permissions for `id-token: write`')
+        }
+      }
+    }
+
+    if (!idToken) {
+      log.silly('oidc', 'Exiting OIDC, no id_token available')
+      return undefined
+    }
+
+    const response = await npmFetch.json(new URL('/-/npm/v1/oidc/token/exchange', registry), {
+      ...opts,
+      method: 'POST',
+      body: JSON.stringify({
+        package_name: packageName,
+        id_token: idToken,
+      }),
+    })
+
+    if (!response?.token) {
+      throw new Error('OIDC token exchange failure: missing token in response body')
+    }
+    const parsedRegistry = new URL(registry)
+    const regKey = `//${parsedRegistry.host}${parsedRegistry.pathname}`
+    const authTokenKey = `${regKey}:_authToken`
+    /*
+     * The "opts" object is a clone of npm.flatOptions and is passed through the `publish` command,
+     * eventually reaching `otplease`. To ensure the token is accessible during the publishing process,
+     * it must be directly attached to the `opts` object.
+     * Additionally, the token is required by the "live" configuration or getters within `config`.
+     */
+    opts[authTokenKey] = response.token
+    config.set(authTokenKey, response.token, 'user')
```

Comment:
This seems like a code smell.  Either we're using config or we're using flatOptions.  It seems we're trying to do both?